src/gt4py/next/iterator/type_system/type_synthesizer.py

```diff
@@ -270,82 +277,196 @@ def apply_lift(
+            return input_
+        case _:
+            raise TypeError(f"Unexpected type: {type(input_)}")
+
+
+def _resolve_dimensions(
+    input_dims: list[common.Dimension],
+    shift_tuple: tuple[itir.OffsetLiteral, ...],
+    offset_provider_type: common.OffsetProviderType,
+) -> list[common.Dimension]:
+    """
+    Resolves the final dimensions by applying shifts from the given shift tuple.
+
+    Args:
+        - input_dims: A list of initial dimensions to resolve.
+        - shift_tuple: A tuple of offset literals defining the shift.
+        - offset_provider_type: Offset provider dictionary.
+
+    Returns:
+        A list of resolved dimensions after applying the shifts.
+    """
+    resolved_dims = []
+    for input_dim in input_dims:
+        for off_literal in reversed(
+            shift_tuple[::2]
+        ):  # Only OffsetLiterals are processed, located at even indices in shift_tuple. Shifts are applied in reverse order: the last shift in the tuple is applied first.
+            offset_type = offset_provider_type[off_literal.value]  # type: ignore [index] # ensured by accessing only every second element
+            if isinstance(offset_type, common.Dimension) and input_dim == offset_type:
+                continue  # No shift applied
+            if isinstance(offset_type, (fbuiltins.FieldOffset, common.NeighborConnectivityType)):
+                if input_dim == offset_type.codomain:  # Check if input fits to offset
+                    input_dim = offset_type.domain[0]  # Update input_dim for next iteration
+        resolved_dims.append(input_dim)
+    return resolved_dims
+
+
 @_register_builtin_type_synthesizer
 def as_fieldop(
     stencil: TypeSynthesizer,
     domain: Optional[it_ts.DomainType] = None,
     *,
     offset_provider_type: common.OffsetProviderType,
 ) -> TypeSynthesizer:
-    # In case we don't have a domain argument to `as_fieldop` we can not infer the exact result
-    # type. In order to still allow some passes which don't need this information to run before the
-    # domain inference, we continue with a dummy domain. One example is the CollapseTuple pass
-    # which only needs information about the structure, e.g. how many tuple elements does this node
-    # have, but not the dimensions of a field.
-    # Note that it might appear as if using the TraceShift pass would allow us to deduce the return
-    # type of `as_fieldop` without a domain, but this is not the case, since we don't have
-    # information on the ordering of dimensions. In this example
-    #   `as_fieldop(it1, it2 -> deref(it1) + deref(it2))(i_field, j_field)`
-    # it is unclear if the result has dimension I, J or J, I.
-    if domain is None:
-        domain = it_ts.DomainType(dims="unknown")
-
     @TypeSynthesizer
-    def applied_as_fieldop(*fields) -> ts.FieldType | ts.DeferredType:
+    def applied_as_fieldop(
+        *fields: ts.TupleType,
+        shift_results: list[set[tuple[itir.OffsetLiteral, ...]]],
```

Comment:
```suggestion
       #: For each stencil parameter all locations it is `deref`ed on see :func:`gt4py.next.iterator.transforms.trace_stencil`.
        shift_results: list[set[tuple[itir.OffsetLiteral, ...]]],
```
How about `shift_sequences_per_param` as a name instead?